[WIP] teller draft

therealssj opened the pull request:
- [x] Add sky scanner
- [x] Refactor scanner's deposit to allow not-processed, rejected and accepted status
- [x] Adapt all data structures to transfer a kitty box.
- [x] Reservations, people reserve a box so that 2 people dont buy the same box. After successful registration they are given a BTC or SKY address
- [x] Integrate with verification service
- [x] Allow partial deposits
- [x] Return reservation payment deadline in response

therealssj committed:
teller draft

mostly broken, sender implementation not touched

128-95 commented on cmd/teller/teller.go:
```diff
@@ -219,21 +215,20 @@ func run() error {
-				log.WithError(err).Errorf("multiplexer.AddScanner of %s failed", scanner.CoinTypeETH)
-				return err
+			if err := multiplexer.AddScanner(scanSkyService, scanner.CoinTypeSKY); err != nil {
+				log.WithError(err).Errorf("multiplexer.AddScanner of %s failed", scanner.CoinTypeSKY)
```
Missing `return err`?

therealssj replied:
ah yes.

evanlinjin replied:
yes.

evanlinjin replied:
yes. If there is error with adding a sky scanner, return that error. Otherwise, return nil (later on).

evanlinjin commented on src/agent/reservation.go:
```diff
@@ -0,0 +1,157 @@
+	// Address where the buyer should send the payment
+	DepositAddress string
+	// The kitty box of this reservation
+	Box *box.Box
```
Just wondering on the thinking for having a `box.Box` type? Aren't all kitties represented by KittyID anyway so we can use that? 

https://github.com/kittycash/wallet/blob/master/src/iko/kitty.go#L10

therealssj replied:
every kitty will be inside a box so just following that analogy
`Box` just contains the `KittyID` only.

evanlinjin replied:
Okay sweet.

evanlinjin commented on src/agent/user.go:
```diff
@@ -0,0 +1,25 @@
+
+
+// User represent a kitty cash user
+type User struct {
```
I need everyone's opinion on this, but I was thinking about having email confirmations for the reservation process. So to initiate reservation, we get:
* What kitty to reserve.
* Payment method.
* Address to have kitty under.
* Captcha code.
* Send them confirmation code via email.
* Get them to input confirmation code.

Hence;
* Each email can only reserve "n" kitties at a time.
* But also, each address can only reserve "n" kitties at one time too.

evanlinjin commented on src/box/box.go:
```diff
@@ -0,0 +1,33 @@
+//@TODO what other info can kitty box store?
+//@TODO implement this
+// Box represents a unique kitty box
+type Box struct {
```
@128-95 will need to work with you for this :+1: as he is implementing the API for determining the kitty details for a given `KittyID`.

This API will interact with DB where it stores such information, where each information is signed under a public/private key pair.

So it'll be essentially a key-value store of;
* Key: `KittyID`
* Value: 

```go
type SignedBoxType struct {
    BoxType
    Sig cipher.Sig
}
```

But teller will still record what kitties are for sale.

Opinions? :1234:

therealssj replied:
so this will be an external api which will contain the signed kitty data?

I guess this is good as boxes are just temp for the iko but the kitty data is going to remain so having it store separately sounds good

128-95 replied:
We will need a boolean in the box to indicates if the box is open or not. Have to think about how we are going to open those boxes.

evanlinjin reviewed (commented):
:100: :+1:

therealssj committed:
sender implementation and implement some checks

therealssj committed:
Add sky scanner tests

therealssj committed:
Add more tests

therealssj committed:
Add pruning to Gopkg.toml

therealssj committed:
Fix http tests

therealssj committed:
Initial attempt at partial deposits

therealssj committed:
Fix race conditions

therealssj committed:
Remove go get -t from travis

therealssj committed:
fix fake btcd

therealssj committed:
Refactor sender and exchange

- Make exchange deposit handling resumable again
- fix tests according to new rpc api changes

therealssj committed:
Fix linting errors

therealssj committed:
Refactor reservation

therealssj committed:
Add kitty api as submodule

therealssj committed:
integrate verification service

therealssj committed:
fix scanner tests

therealssj committed:
fix dependencies

briancaine commented:
@therealssj What's our progress on this? Timeframe?

therealssj commented:
I will have this ready by end of this week

therealssj committed:
unfinished kittyapi integration

therealssj committed:
fixing some linting bugs

therealssj committed:
update kitty api submodule

therealssj committed:
add missing kittyapi address

therealssj committed:
remove kitty api submodule

therealssj committed:
remove otc frontend

therealssj committed:
keep kitty api and teller in sync

therealssj committed:
add user if not already exists

therealssj committed:
refactor addr manager

therealssj committed:
fix getdepositaddress

therealssj committed:
refactor getdepositaddress
